test/unit/org/apache/cassandra/db/guardrails/GuardrailPageSizeTest.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.guardrails;
+
+import java.util.Collections;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.cql3.CQLStatement;
+import org.apache.cassandra.cql3.QueryOptions;
+import org.apache.cassandra.cql3.QueryProcessor;
+import org.apache.cassandra.db.ConsistencyLevel;
+import org.apache.cassandra.service.ClientState;
+import org.apache.cassandra.service.QueryState;
+import org.apache.cassandra.transport.ProtocolVersion;
+
+import static java.lang.String.format;
+
+/**
+ * Tests the guardrail for the page size, {@link Guardrails#pageSize}.
+ */
+public class GuardrailPageSizeTest extends ThresholdTester
+{
+    private static final int PAGE_SIZE_WARN_THRESHOLD = 5;
+    private static final int PAGE_SIZE_ABORT_THRESHOLD = 10;
+
+    public GuardrailPageSizeTest()
+    {
+        super(PAGE_SIZE_WARN_THRESHOLD,
+              PAGE_SIZE_ABORT_THRESHOLD,
+              DatabaseDescriptor.getGuardrailsConfig().getPageSize(),
+              Guardrails::setPageSizeThreshold,
+              Guardrails::getPageSizeWarnThreshold,
+              Guardrails::getPageSizeAbortThreshold);
+    }
+
+    @Before
+    public void setupTest()
+    {
+        createTable("CREATE TABLE IF NOT EXISTS %s (k INT, c INT, v TEXT, PRIMARY KEY(k, c))");
+    }
+
+    @Test
+    public void testSelectStatementAgainstPageSize() throws Throwable
+    {
+        assertPagingValid("SELECT * FROM %s", 3);
+        assertPagingValid("SELECT * FROM %s", PAGE_SIZE_WARN_THRESHOLD);
+
+        assertPagingWarns("SELECT * FROM %s", 6, format("Query for table %s with page size %s exceeds warning threshold of %s.", currentTable(), 6, PAGE_SIZE_WARN_THRESHOLD));
+        assertPagingWarns("SELECT * FROM %s", PAGE_SIZE_ABORT_THRESHOLD, format("Query for table %s with page size %s exceeds warning threshold of %s.", currentTable(), PAGE_SIZE_ABORT_THRESHOLD, PAGE_SIZE_WARN_THRESHOLD));
+
+        assertPagingAborts("SELECT * FROM %s", 11, format("Aborting %s table query, page size %s exceeds abort threshold of %s.", currentTable(), 11, PAGE_SIZE_ABORT_THRESHOLD));
```

Comment:
We could also include an aggregation query (e.g `SELECT COUNT(*)`, and probably also a `LIMIT` clause. The `LIMIT` case is interesting because if it's lesser than the page size the triggering of the guardrail would depend on where in `SelectStatement` we have put the guardrail check. So it could be something like:
```suggestion
        // regular query
        String query = "SELECT * FROM %s";
        assertPagingValid(query, 3);
        assertPagingValid(query, PAGE_SIZE_WARN_THRESHOLD);
        assertPagingWarns(query, 6);
        assertPagingWarns(query, PAGE_SIZE_ABORT_THRESHOLD);
        assertPagingAborts(query, 11);

        // aggregation query
        query = "SELECT COUNT(*) FROM %s WHERE k=0";
        assertPagingValid(query, 3);
        assertPagingValid(query, PAGE_SIZE_WARN_THRESHOLD);
        assertPagingWarns(query, 6);
        assertPagingWarns(query, PAGE_SIZE_ABORT_THRESHOLD);
        assertPagingAborts(query, 11);

        // query with limit over thresholds
        query = "SELECT * FROM %s LIMIT 100";
        assertPagingValid(query, 3);
        assertPagingValid(query, PAGE_SIZE_WARN_THRESHOLD);
        assertPagingWarns(query, 6);
        assertPagingWarns(query, PAGE_SIZE_ABORT_THRESHOLD);
        assertPagingAborts(query, 11);

        // query with limit under thresholds
        query = "SELECT * FROM %s LIMIT 1";
        assertPagingValid(query, 3);
        assertPagingValid(query, PAGE_SIZE_WARN_THRESHOLD);
        assertPagingValid(query, 6);
        assertPagingValid(query, PAGE_SIZE_ABORT_THRESHOLD);
        assertPagingValid(query, 11);
```
Please note that `SELECT * FROM %s LIMIT 1` internally uses a lower page size because of the limit, so the guardrail isn't triggered. I think this behaviour is correct.

Reply:
done

Reply:
```java
        // query with limit under thresholds
        query = "SELECT * FROM %s LIMIT 1";
        assertPagingValid(query, 3);
        assertPagingValid(query, PAGE_SIZE_WARN_THRESHOLD);
        assertPagingValid(query, 6);
        assertPagingValid(query, PAGE_SIZE_ABORT_THRESHOLD);
        assertPagingValid(query, 11);
```
I am looking at this case - and honestly it works like this completely accidentally, I thought execute() covers all statements(so, there will be no query that obey the guardrail) :)

Reply:
Yep, I found it playing with tests. I think that either using strictly the supplied page size or the min between page size and limit are valid options depending on how you look at it, although I think I slightly prefer the current behaviour.